File: studio-idea-plugin/src/main/java/com/evolveum/midpoint/studio/ui/connector/generator/step/basic/CreateConnectorStep.java
package com.evolveum.midpoint.studio.ui.connector.generator.step.basic;

import com.evolveum.midpoint.prism.PrismObject;
import com.evolveum.midpoint.studio.client.AuthenticationException;
import com.evolveum.midpoint.studio.impl.MidPointClient;
import com.evolveum.midpoint.studio.impl.UploadResponse;
import com.evolveum.midpoint.studio.ui.connector.generator.ConnectorGeneratorDataModel;
import com.evolveum.midpoint.studio.ui.connector.generator.ConnectorGeneratorWizard;
import com.evolveum.midpoint.studio.ui.connector.generator.component.GenerateConnectorBadge;
import com.evolveum.midpoint.studio.ui.connector.generator.step.ConnectorGeneratorGeneralWizardStep;
import com.evolveum.midpoint.util.exception.SchemaException;
import com.evolveum.midpoint.xml.ns._public.common.common_3.*;
import com.intellij.openapi.progress.ProgressIndicator;
import com.intellij.openapi.progress.ProgressManager;
import com.intellij.openapi.progress.Task;
import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.awt.*;
import java.io.*;

public class CreateConnectorStep extends ConnectorGeneratorGeneralWizardStep {

    private CreateConnector createConnector;
    private final JPanel mainPanel = new JPanel(new BorderLayout());

    public CreateConnectorStep(
            ConnectorGeneratorWizard wizardContext,
            MidPointClient client,
            ConnectorGeneratorDataModel dataModel,
            GenerateConnectorBadge.State state,
            boolean isHeader
    ) {
        super(wizardContext, client, dataModel, state, isHeader);
        mainPanel.setName("Creating Connector");
    }

    @Override
    public void _init() {
        if (!initialized) {
            initialized = true;

            setState(GenerateConnectorBadge.State.IN_PROGRESS);

            createConnector = new CreateConnector();

            ProgressManager.getInstance().run(new Task.Backgroundable(getClient().getProject(),
                    "CreateConnector submit operation",
                    true
            ) {

                private String token;

                @Override
                public void run(@NotNull ProgressIndicator progressIndicator) {
                    try {
                        token = getClient().submitOperationCreateConnector(getDataModel().connectorDevelopmentType.getOid());
                    } catch (SchemaException | IOException | AuthenticationException e) {
                        throw new RuntimeException(e);
                    }
                }

                @Override
                public void onSuccess() {
                    var statusPanel = createConnector.getStatusPanel();

                    if (token == null) {

                        printAlertPanel(mainPanel, statusPanel, "Error", "Token is Null");
                        return;
                    }

                    printWaitingPanel(
                            mainPanel,
                            statusPanel,
                            "Creating Connector...",
                            """
                                  We use the connector's basic information to create a test instance for development and testing purposes.
                                  """
                    );

                    statusPanel.getElapsedLabel().start();

                    getResult(token).whenComplete((statusInfoResult, ex) -> {
                        if (ex != null) {
                            printAlertPanel(mainPanel, statusPanel, "Error", ex.getMessage());
                        } else {
                            var result = statusInfoResult.getConnDevCreateConnectorResult();

                            if (result != null && result.getConnectorRef() != null) {

                                try {

                                    if (createResourceObject(result.getConnectorRef()) == null) {
                                        throw new Exception("Error creating resource object for connector: " + result.getConnectorRef());
                                    } else {
                                        printAlertPanel(
                                                mainPanel,
                                                statusPanel,
                                                OperationResultStatusType.SUCCESS.name(),
                                                "Successfully Generated Connector Development"
                                        );

                                        canGoNext(true);
                                        getWizardContext().updateWizardButtons();
                                        setState(GenerateConnectorBadge.State.COMPLETE);
                                    }
                                } catch (Exception e) {
                                    printAlertPanel(mainPanel, statusPanel, OperationResultStatusType.UNKNOWN.name(), e.getMessage());
                                }
                            } else {

                                printAlertPanel(
                                        mainPanel,
                                        statusPanel,
                                        OperationResultStatusType.UNKNOWN.name(),
                                        "ConnectorRef Null"
                                );
                            }
                        }

                        statusPanel.getElapsedLabel().stop();
                    });

                    super.onSuccess();
                }
            });
        }

        super._init();
    }

    @Override
    public JComponent getComponent() {
        return mainPanel;
    }

    @Override
    public SmartIntegrationOperationStatusInfoType getStatusInfo(
            String token
    ) throws SchemaException, AuthenticationException, IOException {
        return getClient().getStatusInfoCreateConnector(token);
    }

    private UploadResponse createResourceObject(
            @NotNull ObjectReferenceType objectReferenceType
    ) throws SchemaException, AuthenticationException, IOException {

        var prismContext = getClient().getPrismContext();

        if (prismContext == null) {
            return null;
        }

        PrismObject<ResourceType> resourceObject = prismContext.createObject(ResourceType.class);
        ResourceType resourceType = resourceObject.asObjectable();
        resourceType.setName(getDataModel().connectorDevelopmentType.getName());
        ObjectReferenceType clonedRef = objectReferenceType.clone();
        resourceType.setConnectorRef(clonedRef);

        return getClient().upload(resourceType.asPrismObject(), null);
    }
}

File: studio-idea-plugin/src/main/kotlin/com/evolveum/midpoint/studio/ui/connector/generator/step/basic/CreateConnectorStep.kt
/*
 * Copyright (C) 2010-2026 Evolveum and contributors
 *
 * Licensed under the EUPL-1.2 or later.
 */

package com.evolveum.midpoint.studio.ui.connector.generator.step.basic

import com.amazon.ion.NullValueException
import com.evolveum.midpoint.prism.PrismConstants
import com.evolveum.midpoint.prism.PrismContext
import com.evolveum.midpoint.studio.client.AuthenticationException
import com.evolveum.midpoint.studio.client.RPCOperation
import com.evolveum.midpoint.studio.impl.MidPointClient
import com.evolveum.midpoint.studio.impl.UploadResponse
import com.evolveum.midpoint.studio.ui.connector.generator.ConnectorGeneratorWizard
import com.evolveum.midpoint.studio.ui.connector.generator.component.GenerateConnectorBadge
import com.evolveum.midpoint.studio.ui.connector.generator.component.StatusPanel
import com.evolveum.midpoint.studio.ui.connector.generator.step.ConnectorGeneratorGeneralWizardStep
import com.evolveum.midpoint.util.exception.SchemaException
import com.evolveum.midpoint.xml.ns._public.common.common_3.ObjectReferenceType
import com.evolveum.midpoint.xml.ns._public.common.common_3.OperationResultStatusType
import com.evolveum.midpoint.xml.ns._public.common.common_3.ResourceType
import com.evolveum.prism.xml.ns._public.types_3.PolyStringType
import com.intellij.openapi.ui.DialogPanel
import com.intellij.util.concurrency.EdtExecutorService
import java.io.IOException

class CreateConnectorStep (
    wizardContext : ConnectorGeneratorWizard,
    client : MidPointClient,
    state : GenerateConnectorBadge.State,
    isHeader : Boolean
) : ConnectorGeneratorGeneralWizardStep(wizardContext, client, state, isHeader) {

    override val dialogPanel: DialogPanel by lazy {
        createDialogPanel(
            "Creating Connector"
        ) { }
    }

    override val nextButtonText: String
        get() = "Download Connector"

    override fun _init() {
        super._init()

        if (!existConnDev()) return

        if (!wizardContext.isLeavingStepByPreviousTouch && dataModel.occurredChanges) {
            submitOperation()
        }
    }

    private fun submitOperation() {
        replaceContent(
            getLoadingComponent(
                statusPanel,
                "Creating Connector...",
                """
                We use the connector's basic information to create a test instance for development and testing purposes.
                """
            )
        )

        submitOperation(
            {
                client.submitOperationConnGenerator(
                    RPCOperation.RPC_CREATE_CONNECTOR,
                    mapOf("oid" to dataModel.connectorDevelopment.oid),
                    null
                )
            },
            { token ->
                client.getStatusInfoConnGenerator(
                    RPCOperation.RPC_CREATE_CONNECTOR,
                    token
                )
            },
            client.project,
            "Create Connector submit operation",
            true
        ).thenAcceptAsync(
            { statusInfo ->
                if (statusInfo.status == OperationResultStatusType.SUCCESS) {
                    try {
                        val result = statusInfo.result.connDevCreateConnectorResult
                        val connectorRef = result?.connectorRef ?: run {
                            throw NullValueException("ConnectorRef Null")
                        }

                        val testingResource = createTestingResource(connectorRef) ?: run {
                            throw NullValueException("Failed to create testing resource")
                        }

                        val resourceRef = ObjectReferenceType()
                        resourceRef.oid = testingResource.oid
                        resourceRef.type = ResourceType.COMPLEX_TYPE
                        resourceRef.relation = PrismConstants.Q_ANY

                        dataModel.connectorDevelopment.testing.testingResource = resourceRef
                        dataModel.connectorDevelopment.connector.connectorRef = connectorRef

                        upsertConnectorDevelopmentType(dataModel.connectorDevelopment)

                        statusPanel.status = StatusPanel.Status.SUCCESS
                        replaceContent(
                            getAlertComponent(
                                statusPanel,
                                statusInfo.status.name,
                                "Successfully Generated Connector Development"
                            )
                        )

                        state = GenerateConnectorBadge.State.COMPLETE
                        canGoNext = true
                    } catch (e: Exception) {
                        statusPanel.status = StatusPanel.Status.ERROR
                        replaceContent(
                            getAlertComponent(
                                statusPanel,
                                statusPanel.status?.name ?: "",
                                e.message
                            )
                        )
                    }
                } else {
                    statusPanel.status = StatusPanel.Status.ERROR
                    replaceContent(
                        getAlertComponent(
                            statusPanel,
                            statusPanel.status?.name ?: "",
                            statusInfo.message
                        )
                    )
                }

                statusPanel.elapsedLabel?.stop()
            },
            EdtExecutorService.getInstance()
        ).whenCompleteAsync(
            { _, throwable ->
                if (throwable != null) {
                    statusPanel.status = StatusPanel.Status.ERROR
                    replaceContent(
                        getAlertComponent(
                            statusPanel,
                            statusPanel.status?.name ?: "",
                            throwable.localizedMessage
                        )
                    )

                    statusPanel.elapsedLabel?.stop()
                }
            },
            EdtExecutorService.getInstance()
        )
    }

    @Throws(SchemaException::class, AuthenticationException::class, IOException::class)
    private fun createTestingResource(
        connectorRef: ObjectReferenceType
    ): ResourceType? {
        val prismContext: PrismContext = client.prismContext ?: return null
        val resourceObject = prismContext.createObject(ResourceType::class.java)
        val resourceType = resourceObject.asObjectable()
        resourceType.name = PolyStringType("Resource - " + dataModel.connectorDevelopment.name?.orig)
        resourceType.connectorRef = connectorRef.clone()

        return client.upsert(resourceType.asPrismObject(), null)
    }
}